### PR TITLE
Google Product Search&Vision API

### LINKED ISSUE #82: Implement the product search&vision functionality of the API and create basic front-end to demonstrate
We add get similar products, create product set, add and delete products, add reference image, update a product and filter products by category functions to our api and create basic front-end to demonstrate.

### PR DESCRIPTION
**Pull request by asliayk**
@karalleyna and I have been working on the Google product search API. 

Our API consists of products and product sets. We've thought each brand(vendor) has its own set and customers should be able to reach these sets to see products. When the user comes to the homepage, s/he sees 11 options: 5 for customers, 6 for vendors. Here are the brief explanations for each option:

Vendor options:

- Create product set: Vendor enters set id and set name and if a set with that
id doesn't exist, a new set will be created.   
- Delete product set: Vendor enters a set id and if a set with that id exists, it will
be deleted.    
- Add product to a product set: The vendor enters product id, set id, product name, product category, product description, color and style(valid styles: home, women, men, children) to add that product to that product set.     
- Update product: The vendor enters product id, key name(“color” or “style”) and value for that key of the product to update the key value. If the key name is style but value is not equal to home, children, women or men, it warns to user to enter a valid key value.
- Delete product: The vendor enters the id of the product s/he wants to delete and product will be deleted.
- Add reference image for a product: The vendor add a picture from google cloud storage for a product by entering product id and reference image uri. S/he can add more than one picture for a product. All pictures are stored in Google cloud storage. 

Customer options: 

- List all products: Customer clicks list button to see the list of all products, the list
will be shown on a new tab.    
Output: 

listallproducts

- List sets: Customer clicks list button to see the list of all sets, the list
will be shown on a new tab.   
Output:

listsets

- List products in product set: Customer enters a set id and clicks list button 
to see the list of all products in the set with that id. If a set with that id does not
exist or it is empty, it warns the user. If it exists and contains products, these 
products will be listed on a new tab.   
Output:

listsproductsinset

- Filter products by category in a product set: Customer enters a set id and category(valid 
categories: apparel, apparel-v2, homegoods, toys, homegoods-v2 and packagedgoods-v1) 
and the products of that category in that set are listed on a new tab. If no product 
exists in that category or the set is empty/does not exist, it warns the user.   
Output:

filter

- List similar products: The customer can see whether there exists a similar product for a product that s/he wants to buy or look for by giving a product picture url(file path) from his/her computer, its category name, filter (ex. style=men), and list for similar products and similarity scores will be shown on a new tab as a sorted list by their scores.  
Output:

listsimilar2

### CODE REVIEW

**Comment by karalleyna** [comment]
The main purpose in choosing this **Google Product Search API** with @asliayk is implementing list similar products function of the image client of Google Product Search API in order to use it in the future process of Getflix. For example, a customer wants to buy a dress with specific features: its color, its style etc. She finds some dress pictures from the internet which are similar to her dream dress on her mind. However, she doesn’t have much time and doesn’t want to look all products one by one on an e-commerce web site to find some similar dresses. With this API feature, she just enters the file path of the picture and the similar products in Getflix are listed with just one click.
Other methods such as creating product set, updating product and deleting product enable the necessary environment which list similar product function uses. These methods use **GET, POST,PATCH** and **DELETE** HTTP requests rather than stuck with just **GET** request.
- Initially, we created a product set named cmpe352 consisting of dresses with different types and colors. 
productCatalog1
productCatalog2
productCatalog3

- If you want to use this product set, it is enough to give the file path of the dress image you choose.  The results will be as follows:
  - **The dress which we have searched:**
test_dress
  - **Output**
output

- If you want to add product and its picture, just make sure that the pictures are stored on a bucket in google cloud storage and publicly readable. You should enter their google storage uri e.g.  gs://cmpe352_example/black_long_dress.jpg.

File: practice_app/package.json [hunk 1]
```diff
@@ -8,11 +8,14 @@
         "start": "next start",
         "test": "jest --passWithNoTests",
         "dev:test": "jest --passWithNoTests --watch",
-        "util:prettier": "prettier --write package.json \"./pages/**/*.js\"",
+        "util:prettier": "prettier --write package.json \"./pages/*/.js\"",
         "util:lint-staged": "lint-staged",
-        "util:eslint": "eslint --fix ./pages/**/*.js",
+        "util:eslint": "eslint --fix ./pages/*/.js",
```
Comment (MehdiSaffar) [inline practice_app/package.json]: Did you guys change this? If so, why?

File: practice_app/pages/api/getAllProducts.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+          mytext += "<li>" + product.name + "</li>";
+        }
+
+        // Sends response text to HTML page
+        if (mytext !== "<h1> Products </h1>") {
+          res.on("finish", resolve);
+          res.send("<p>" + mytext + "</p>");
+        }
+      })
+      .catch((err) => {
```
Comment (MehdiSaffar) [inline practice_app/pages/api/getAllProducts.js]: The backend shouldn't return raw HTML. I would suggest returning the a JSON for the data of the products and doing the rendering on the frontend.

File: practice_app/pages/api/getAllProducts.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
+const getAllProducts = (req, res) =>
+  new Promise(async (resolve) => {
+    // Imports the Google Cloud client library
```
Comment (MehdiSaffar) [inline practice_app/pages/api/getAllProducts.js]: It is enough to do `const getAllProduct = async (req, res) =>` without any explicit Promise as an `async` function is a function that returns a `Promise itself`. Also remove any `res.on('finish', resolve)` as it is not needed and simply use the regular `res.json({...})`

File: practice_app/pages/api/getAllProducts.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    // Google cloud project id and region name
+    const projectId = "brilliant-era-276800";
+    const location = "europe-west1";
+
+    // Resource path that represents Google Cloud Platform location.
+    const formattedParent = client.locationPath(projectId, location);
+
+    // Initializes HTML text
+    var mytext = "<h1> Products </h1>";
```
Comment (MehdiSaffar) [inline practice_app/pages/api/getAllProducts.js]: `var` is old javascript. Use `const` for immutable variables, `let` for mutable variables.

File: practice_app/pages/api/getAllProducts.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
+const getAllProducts = (req, res) =>
+  new Promise(async (resolve) => {
+    // Imports the Google Cloud client library
+    const vision = require("@google-cloud/vision");
```
Comment (MehdiSaffar) [inline practice_app/pages/api/getAllProducts.js]: It is bad practice to require inside functions, you can move any require to the top level (outside of the function)

File: practice_app/pages/api/getProductCatalog.js [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+    const location = "europe-west1";
+    // Method of the request
+    const method = req.method;
+
+    switch (method) {
+      case "POST": {
+        const requestType = req.body.type;
+        // Creates a product and then adds it to the product set
+        if (requestType == "product") {
+          console.log("aaaaa");
```
Comment (MehdiSaffar) [inline practice_app/pages/api/getProductCatalog.js]: Please remove such console logs.

File: practice_app/pages/api/getProductCatalog.js [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+          };
+
+          // Initializes the request object
+          const createRequest = {
+            parent: locationPath,
+            product: product,
+            productId: productId,
+          };
+          // Creates the product via 'createProduct' method of the Google Product Search API
+          await client.createProduct(createRequest).catch((err) => {
```
Comment (MehdiSaffar) [inline practice_app/pages/api/getProductCatalog.js]: I would suggest not mixing `async ... await` syntax with `.then(..).catch(...)`

https://css-tricks.com/understanding-async-await/

File: practice_app/pages/api/getProductCatalog.js [hunk 8]
```diff
@@ -0,0 +1,278 @@
+const getProductCatalog = (req, res) =>
```
Comment (MehdiSaffar) [inline practice_app/pages/api/getProductCatalog.js]: This function is too long. Please break it down to different functions based on request type.

File: practice_app/pages/api/getProductCatalog.test.js [hunk 9]
```diff
@@ -0,0 +1,341 @@
+const getProductSet = async (req) => {
```
Comment (MehdiSaffar) [inline practice_app/pages/api/getProductCatalog.test.js]: This function is too long.

Also is this the actual implementation? It doesn't belong to a test file.

File: practice_app/pages/api/getProductsByCategory.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    const projectId = "brilliant-era-276800";
+    const location = "europe-west1";
+
+    // takes product set id and category name from the request url
+    var str = request.url;
+    var n = str.indexOf("=");
+    var n1 = str.indexOf("&");
+    var first = str.substring(n + 1, n1);
+    var n2 = str.indexOf("=", n + 1);
+    var second = str.substring(n2 + 1, str.length);
```
Comment (MehdiSaffar) [inline practice_app/pages/api/getProductsByCategory.js]: You can use const here

File: practice_app/pages/api/getProductsInSet.js [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+            const images = responses[0];
+          for (const image of images) {
+           console.log(product.name);   
+           console.log(image.uri); 
+          
+          }).catch(err => {
+            console.error(err);
+          }); */
+        }
+
```
Comment (MehdiSaffar) [inline practice_app/pages/api/getProductsInSet.js]: If this code is no longer necessary please get rid of this comment.

File: practice_app/pages/api/getSets.js [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+  new Promise(async (resolve) => {
+    // Imports the Google Cloud client library
+    const vision = require("@google-cloud/vision");
+
+    // Creates a client
+    const client = new vision.ProductSearchClient();
+
+    // Google cloud project id and region name
+    const projectId = "brilliant-era-276800";
+    const location = "europe-west1";
```
Comment (MehdiSaffar) [inline practice_app/pages/api/getSets.js]: These constants appear in multiple places. This is a good fit for environment variables.

Comment (MehdiSaffar) [reply practice_app/pages/api/getSets.js]: Or maybe a simple file `config.json` that is imported?

File: practice_app/pages/api/getSets.js [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        // sending response text to HTML page
+        if (mytext !== "<h1> Sets </h1>") {
+          response.on("finish", resolve);
+          response.send("<p>" + mytext + "</p>");
+        }
+      })
+      .catch((err) => {
+        // shows error message
+        console.error(err);
+      });
```
Comment (MehdiSaffar) [inline practice_app/pages/api/getSets.js]: An error should be sent back as a response or the request hangs.

File: practice_app/pages/api/getSimilarProducts.js [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        keyValue[1] = keyValue[1].split("%3A").join(":");
+      }
+      if (keyValue[1].includes("%2F")) {
+        keyValue[1] = keyValue[1].split("%2F").join("/");
+      }
+      if (keyValue[1].includes("%5C")) {
+        keyValue[1] = keyValue[1].split("%5C").join("\\");
+      }
+      queryParams[keyValue[0]] = keyValue[1];
+    });
```
Comment (MehdiSaffar) [inline practice_app/pages/api/getSimilarProducts.js]: Please do not reinvent the wheel by trying to parse url query params yourself. This is brittle code. There are already libraries out there that do this. Try https://www.npmjs.com/package/query-string

File: practice_app/pages/api/getSimilarProducts.js [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    results.forEach((result) => {
+      temp +=
+        "<h2>Product id:" + result["product"].name.split("/").pop(-1) + "</h2>";
+      temp += "<li>Product name: " + result["product"].displayName + "</li>";
+      temp +=
+        "<li>Product description: " + result["product"].description + "</li>";
+      temp +=
+        "<li>Product category: " + result["product"].productCategory + "</li>";
+      temp += "<li>Product similarity score: " + result.score + "</li>";
+    });
```
Comment (MehdiSaffar) [inline practice_app/pages/api/getSimilarProducts.js]: Again please do not render like this. Use the client-side react code for proper rendering.

File: practice_app/pages/index.js [hunk 16]
```diff
@@ -1,177 +1,735 @@
...
+    if (
+      !(
+        document.getElementById("category-title").value == "apparel" ||
+        document.getElementById("category-title").value == "homegoods" ||
+        document.getElementById("category-title").value == "toys" ||
+        document.getElementById("category-title").value == "apparel-v2" ||
+        document.getElementById("category-title").value == "homegoods-v2" ||
+        document.getElementById("category-title").value == "toys-v2" ||
+        document.getElementById("category-title").value == "packagedgoods-v1"
+      )
```
Comment (MehdiSaffar) [inline practice_app/pages/index.js]: Shorter check is 
```js
['apparel', 'homegoods', ...,  'packagedgoods-v1'].includes(document.getElementById("category-title").value)
```

File: practice_app/pages/index.js [hunk 17]
```diff
@@ -1,177 +1,735 @@
...
+        <label>
+          SetID:
+          <input
+            type="text"
+            placeholder="Add set id"
+            id="product-set-title"
+            required
+          />
+        </label>
+      </form>
```
Comment (MehdiSaffar) [inline practice_app/pages/index.js]: The HTML `<form>` tag is used to wrap a whole form, not a form element. It should wrap the entire form.

**Review by MehdiSaffar — CHANGES_REQUESTED** [review]
I like the feature and I appreciate how much thought you put into it and how well you description is put out with images and test reports and so on. Good job :) 

I haven't tested the app yet as I am having trouble setting it up.
There are merge conflicts with master branch. Make sure to merge master to your branch, do the necessary code changes and push again.

**Review by omerfdeniz — APPROVED** [review]
Your code is clean and easy-to-read with explanatory comments. I see that you paid attention to not making http requests in unit tests, which is important for preventing future unexplainable errors. You should also remove console logs created for testing purposes before merge. All in all, I found your work pretty useful and demanding. The functionalities you provide with this api is essential for our product. If you can make the stated improvements to the code, it will be better in the long run, but even right now code is working properly. Excellent job!

**Comment by asliayk** [comment]
- [x]	Unnecessary comments and console.log’s are removed, var is changed to let.
- [x]	We used shorter checks with includes.
- [x]	We broke our main function(getProductCatalog) into smaller functions based on their types(create, delete or update)
- [x]	We removed res.on and used res.json instead.

**Comment by karalleyna** [comment]
- [x]	We wrote our frontend part as a separate js file and stored in components folder. (SimilarProduct.js)
- [x]	Since we updated our code, we recreated our unit tests.
- [x]	We updated our listing similar products feature, similar products are shown on the same page, not on a new tab now. 
- [x]	We updated GetSimilarProducts.js file with the help of @MehdiSaffar , user can also choose a picture on his/her computer now, s/he does not need to give its file path anymore.

**Comment by MehdiSaffar** [comment]
I merged master correctly by taking the most up to date version of it.
I refactored a few things.

I will accept this pull request. 

Good job guys :)

**Pull request merged**